compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp

```diff
@@ -368,6 +405,150 @@ void LinalgFusePass::runOnOperation() {
     return signalPassFailure();
   }
 }
+namespace {
+/// Pattern to wrap splitReduction transformation.
+struct CodegenSplitReduction
+    : public OpInterfaceRewritePattern<linalg::LinalgOp> {
+  CodegenSplitReduction(MLIRContext *context, bool reductionReordering,
+                        int64_t size, linalg::LinalgTransformationFilter filter,
+                        PatternBenefit benefit = 1)
+      : OpInterfaceRewritePattern<linalg::LinalgOp>(context, benefit),
+        reductionReordering(reductionReordering),
+        size(size),
+        filter(std::move(filter)) {}
+
+  LogicalResult matchAndRewrite(linalg::LinalgOp op,
+                                PatternRewriter &rewriter) const override {
+    // Make sure that
+    // - the pass has not been applied before
+    // - has tensor semantics
+    // - number of reduction loops == 1
+    // - has exactly 1 output
+    // - index map has only projected permutations
+    // - is a linalg generic op
+    // - has exactly 1 input
+    // - if enableReductionReordering is not set, then operand is an int
+    // - innermost dimension of the input operand is reduction
+    // TODO: support named ops, numInputs > 1, and modify lastDim check below
+    // accordingly. If reductionReordering is not enabled by default, it must be
+    // an integer or index type to proceed to allow associative reordering.
+    if (failed(filter.checkAndNotify(rewriter, op))) {
+      return rewriter.notifyMatchFailure(op, "pass has been applied before");
+    }
+    if (!op.hasTensorSemantics()) {
+      return rewriter.notifyMatchFailure(op, "doesn't have tensor semantics");
+    }
+    if (op.getNumReductionLoops() != 1) {
+      return rewriter.notifyMatchFailure(op, "number of reduction loops != 1");
+    }
+    if (op.getNumOutputs() != 1) {
+      return rewriter.notifyMatchFailure(op, "doesn't have exactly 1 output");
+    }
+    if (!op.hasOnlyProjectedPermutations()) {
+      return rewriter.notifyMatchFailure(
+          op, "index map doesn't have only projected permutations");
+    }
```

Comment:
I don't really understand why we can't reuse the upstream preconditions in SplitReduction here, and have to do same checks in advance. E.g., hasOnlyProjectedPermutations looks like the needs from SplitReduction pass, not IREE specific reason. Maybe there are deep reasons that I could not figure out. I'm fine with change, and will leave the decision to you.

Reply:
That is perhaps the only check that is necessary for `splitReduction` to work. Might as well perform the check here instead of creating a precondition checker in llvm.

Reply:
The `hasTensorSemantics`, `getNumReductionLoops`, `getNumOutputs` are all requirements from upstream SplitReduction. filter stuff is fine because that's the way to drive CodegenDriver.